iroh-bytes/src/util/progress.rs

```diff
@@ -121,6 +122,107 @@ pub trait ProgressSender: std::fmt::Debug + Clone + Send + Sync + 'static {
     ) -> WithFilterMap<Self, U, F> {
         WithFilterMap(self, f, PhantomData)
     }
+
+    /// Create a boxed progress sender to get rid of the concrete type.
+    fn boxed(self) -> BoxedProgressSender<Self::Msg>
+    where
+        Self: IdGenerator,
+    {
+        BoxedProgressSender(Arc::new(BoxableProgressSenderWrapper(self)))
+    }
+}
+
+/// A boxed progress sender
+pub struct BoxedProgressSender<T>(Arc<dyn BoxableProgressSender<T>>);
+
+impl<T> Clone for BoxedProgressSender<T> {
+    fn clone(&self) -> Self {
+        Self(self.0.clone())
+    }
+}
+
+impl<T> std::fmt::Debug for BoxedProgressSender<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("BoxedProgressSender").field(&self.0).finish()
+    }
+}
+
+/// Boxable progress sender
+trait BoxableProgressSender<T>: IdGenerator + std::fmt::Debug + Send + Sync + 'static {
```

Comment:
Why do we need this new trait? Can't we just implement `ProgressSender` on `BoxedProgressSender`?

Reply:
I need something dynable inside the BoxedProgressSender. And ProgressSender is not dynable, "can not be made into an object" bla bla bla.

I think the reasons are:

- It has msg as an associated type, not as a type parameter
- send returns an impl Future
- the with_filter and with_filter_map also are not good

I guess you could solve this by always returning a BoxedProgressSender from with_filter and with_filter_map, always returning a boxed future from send, and making the associated type a type parameter. But that would suck - way too much boxing in cases you don't need it...